projects/graph/graph.py

```diff
@@ -11,45 +11,142 @@ def add_vertex(self, vertex):
         """
         Add a vertex to the graph.
         """
-        pass  # TODO
+        self.vertices[vertex] = set()
     def add_edge(self, v1, v2):
         """
         Add a directed edge to the graph.
         """
-        pass  # TODO
+        if v1 in self.vertices and v2 in self.vertices :
+            self.vertices[v1].add(v2)
+        else:
+            raise IndexError("Either {v1} or {v2} vertices does no exist.")
+
     def bft(self, starting_vertex):
         """
         Print each vertex in breadth-first order
         beginning from starting_vertex.
         """
-        pass  # TODO
+        # Create an empty set to store visited nodes
+        visited = set()
+        # Create an empty Queue and enqueue the starting vertex
+        q = Queue()
+        q.enqueue(starting_vertex)
+        # While the queue is not empty...
+        while q.size() > 0:
+            # Dequeue the first vertex
+            v = q.dequeue()
+            # If that vertex has not been visited...
+            if v not in visited:
+                # Mark it as visited
+                visited.add(v)
+                print(f'bft print {v}')
+                # Then add all of its neighbors to the back of the queue
+                for neighbor in self.vertices[v]:
+                    q.enqueue(neighbor)
+
     def dft(self, starting_vertex):
         """
         Print each vertex in depth-first order
         beginning from starting_vertex.
         """
-        pass  # TODO
+        visited = set()
+        s = Stack()
+        s.push(starting_vertex)
+        while s.size() > 0 :
+            v = s.pop()
+            if v not in visited:
+                visited.add(v)
+                print(f'dft print {v}')
+                for neighbor in self.vertices[v]:
+                    s.push(neighbor)
+
```

Comment:
very nice; Only suggestion is to try and use more descriptive variable names rather than (s and v)